design-documents/service-isolation/caching-layer.md

```diff
@@ -0,0 +1,105 @@
+ * Protection against outages - can optionally serve stale content when there is a problem with origin server
+ * Scalability - number of caching nodes can be increased
+ * Flexibility – Varnish Configuration Language (VCL) builds customized solutions, rules and modules
+ * Shielding - the shield node is the only node to get the content from origin server, other cache nodes will fetch from the shield node
+ 
+ ![Public content caching](img/public-cache.jpg)
+
+A private or visitor-specific content should be stored on client-side (browser). After POST requests the cache 
+in the web browser will be flushed and an AJAX call will be done to fetch an updated copy of the private content.
+Also private content cache on client-side should expire according to TTL.
+
+ ![Private content caching](img/private-cache.jpg)
+ 
+ In general, the same caching approach works in current monolith Magento architecture.
+ For service oriented architecture next issues should be solved:
+  * The http-response header should include a set of tags for effective cache purging on reverse-proxy.
+ The set should be created while request is going via the chain of services. Appropriate design should be prepared.
+  * Caching on reverse-proxy should be aligned with request authorization approach. The presence of an authorization token should be taken into account.
+  
+
+### Application data caching
+
+Service application needs to cache such data as DB query results, merged configurations for efficiently and rapidly respond to requests.
+
+The three possible types of caches are the following:
+ * Local caches
+ * Remote caches dedicated to each service
+ * Remote caches shared across the services
+ 
+#### Local caches
+ - Pros: fast, no network traffic associated with retrieving data
+ - Cons: 
+   - shares sources with application
+   - can lead to data inconsistency when multiple instances of application servers are created
+   - the information stored within an individual cache node cannot be shared with other application servers
+   - not follow the 12-factor principles (any data that needs to persist must be stored in a stateful backing service)
+ 
+ ![Local cache](img/local-app-cache.png)
+ 
+#### Remote caches dedicated to each service
+ - Pros: 
+   - scalable, adheres to 12-factor principles
+   - externalize application state
+   - possibility dynamically change the number of service instances
+   - contains only dedicated service data
+ - Cons: 
+   - a slight delay in network calls
+   - difficulties with centralized cache invalidation
+ 
+  ![Remote cache dedicated to each service](img/remote-dedicated-cache.png)
+ 
+#### Remote caches shared across the services
+ - Pros: 
+   - centralized cache invalidation is easy
+   - scalable
+   - adheres to 12-factor principles
+ - Cons: 
+   - cache nodes contain mixed data from different services
+ 
+ ![Remote cache shared across the services](img/remote-cache-cluster.png)
```

Comment:
Are Service 1 and Service 2 different services or these are instances of the same service? If first, this contradicts best practices. Please clarify.

Reply:
> Are Service 1 and Service 2 different services or these are instances of the same service? If first, this contradicts best practices. Please clarify.

On this diagram Service 1 and Service 2 are different services. It is only one of the possible variants and we decided that it's not the best option.

> For application cache I would propose each service has it's own cache this cache should not be shared with other services. If there are multiple instances of the same service, they will use the same cache.

So actually this is model "Remote caches dedicated to each service" that we agreed to use. I'll add statement that  multiple instances of the same service should use the same cache instance.

 > I also propose to use reverse proxy to cache GET requests that don't contain sensitive data.

So you propose to set reverse proxy in front of each service? In that case is there sense of setting reverse proxy in front of BFF?